basin3d/core/monitor.py

```diff
@@ -0,0 +1,228 @@
+"""
+`basin3d.core.monitor`
+**********************
+
+.. currentmodule:: basin3d.core.monitor
+
+:synopsis: The BASIN-3D monitoring module
+:module author: Val Hendrix <vhendrix@lbl.gov>
+
+This module holds the logging functionality of BASIN-3D. It supports
+logging the messages and their contexts
+
+.. contents:: Contents
+    :local:
+    :backlinks: top
+
+"""
+import contextvars
+import logging
+import os
+import uuid
+from functools import wraps
+from logging import config
+from typing import Any, Callable, Dict, List, Optional, Union
+
+import yaml
+
+LOGGER_NAME = __name__
+BASE_PATH = os.path.dirname(__file__)
+LOG_CONFIG_PATH = os.path.join(BASE_PATH, "logging.yaml")
+
+#: A unique identifier for a single request context. Use this
+#: If you want to identify log records from a single context
+#: This is a Context variable which is natively supported in asyncio
+#: and are ready to be used without any extra configuration.
+synthesis_id: contextvars.ContextVar = contextvars.ContextVar('synthesis_id')
+
+#: A unique identifier for a  basin3d context. Use this
+#: If you want to identify log records from a single context
+#: This is a Context variable which is natively supported in asyncio
+#: and are ready to be used without any extra configuration.
+basin3d_where: contextvars.ContextVar = contextvars.ContextVar('basin3d_where')
+
+
+class Basin3dLogger(logging.Logger):
+    """
+    Custom logger for adding realtime context information (basin3d_where)
+    """
+
+    def info(self, msg, *args, **kwargs):
+        kwargs = self._add_extra(**kwargs)
+        super().info(msg, *args, **kwargs)
+
+    def error(self, msg, *args, **kwargs):
```

Comment:
should critical be added? I see it in the MonitorMixin and there are the critical monitoring levels is used in core/catalog.py.

Reply:
Good catch. We can add it when we implement some critical messages.

Reply:
critical messages are used in the catalog.py module already.

Reply:
oh. ok.